In [1]:
from google.colab import files

uploaded = files.upload()

Saving datasetZip.zip to datasetZip.zip


In [2]:
from zipfile import ZipFile

with ZipFile("datasetZip.zip", "r") as zip_ref:
    zip_ref.extractall()

print("Dataset Extracted!")

Dataset Extracted!


In [3]:
import os
for split in ["train","val","test"]:
    print(split)
    for cls in ["real","screen"]:
        folder=os.path.join("dataset",split,cls)
        print(cls, len(os.listdir(folder)))

train
real 45
screen 45
val
real 6
screen 6
test
real 6
screen 6


In [4]:
!pip install -q ultralytics

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.8/41.8 kB 1.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 31.0 MB/s eta 0:00:00


In [5]:
from ultralytics import YOLO

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.


In [6]:
model = YOLO("yolov8n-cls.pt")

In [7]:
results = model.train(

    data="dataset",

    epochs=50,

    imgsz=224,

    batch=8,

    lr0=0.001,

    optimizer="Adam",

    patience=10,

    project="ScreenDetector",

    name="YOLOv8n",

    pretrained=True,

    verbose=True
)

Ultralytics 8.4.83 🚀 Python-3.12.13 torch-2.11.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=8, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=dataset, degrees=0.0, deterministic=True, device=None, dfl=1.5, dis=6.0, distill_model=None, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=224, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.001, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n-cls.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=YOLOv8n, nbs=64, nms=False, opset=None, optimize=False, optimizer=Adam, overlap_mask=True, patience=10, perspective

In [8]:
metrics = model.val()

print(metrics)

Ultralytics 8.4.83 🚀 Python-3.12.13 torch-2.11.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLOv8n-cls summary (fused): 30 layers, 1,437,442 parameters, 0 gradients, 3.3 GFLOPs
train: /content/dataset/train... found 90 images in 2 classes ✅ 
val: /content/dataset/val... found 12 images in 2 classes ✅ 
test: /content/dataset/test... found 12 images in 2 classes ✅ 
val: Fast image access ✅ (ping: 0.1±0.2 ms, read: 2138.1±1509.6 MB/s, size: 673.7 KB)
val: Scanning /content/dataset/val... 12 images, 0 corrupt: 100% ━━━━━━━━━━━━ 12/12 3.9Mit/s 0.0s
               classes   top1_acc   top5_acc: 100% ━━━━━━━━━━━━ 1/1 1.6s/it 1.6s
                   all          1          1
Speed: 0.1ms preprocess, 6.2ms inference, 0.0ms loss, 0.0ms postprocess per image
Results saved to /content/runs/classify/val
ultralytics.utils.metrics.ClassifyMetrics object with attributes:

confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x79e4247c80b0>
curves: []
curves_results: []
fitness: 1.0
keys: ['m

Checking Model on Unseen data of YOLO

In [9]:
from ultralytics import YOLO

model = YOLO("/content/runs/classify/ScreenDetector/YOLOv8n/weights/best.pt")

metrics = model.val(
    data="/content/dataset",
    split="test"
)

print(metrics)

Ultralytics 8.4.83 🚀 Python-3.12.13 torch-2.11.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLOv8n-cls summary (fused): 30 layers, 1,437,442 parameters, 0 gradients, 3.3 GFLOPs
train: /content/dataset/train... found 90 images in 2 classes ✅ 
val: /content/dataset/val... found 12 images in 2 classes ✅ 
test: /content/dataset/test... found 12 images in 2 classes ✅ 
test: Fast image access ✅ (ping: 0.0±0.0 ms, read: 872.3±345.0 MB/s, size: 80.8 KB)
test: Scanning /content/dataset/test... 12 images, 0 corrupt: 100% ━━━━━━━━━━━━ 12/12 3.6Kit/s 0.0s
test: New cache created: /content/dataset/test.cache
               classes   top1_acc   top5_acc: 100% ━━━━━━━━━━━━ 1/1 1.4s/it 1.4s
                   all      0.917          1
Speed: 0.1ms preprocess, 1.4ms inference, 0.0ms loss, 0.0ms postprocess per image
Results saved to /content/runs/classify/val-2
ultralytics.utils.metrics.ClassifyMetrics object with attributes:

confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x79e41052acc

Data Augmentation to improve test accuracy


In [10]:
results2 = model.train(
    data="dataset",
    epochs=80,
    imgsz=224,
    batch=8,

    optimizer="AdamW",
    lr0=5e-4,

    fliplr=0.5,
    flipud=0.0,

    degrees=3,
    translate=0.02,
    scale=0.05,

    hsv_h=0.005,
    hsv_s=0.05,
    hsv_v=0.05,

    mosaic=0.0,
    mixup=0.0,
    erasing=0.05,

    patience=20
)

Ultralytics 8.4.83 🚀 Python-3.12.13 torch-2.11.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=8, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=dataset, degrees=3, deterministic=True, device=None, dfl=1.5, dis=6.0, distill_model=None, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=80, erasing=0.05, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, hsv_h=0.005, hsv_s=0.05, hsv_v=0.05, imgsz=224, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.0005, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=/content/runs/classify/ScreenDetector/YOLOv8n/weights/best.pt, momentum=0.937, mosaic=0.0, multi_scale=0.0, name=train, nbs=64, nms=False, opset=None, optimize=False, optimizer=Ad

In [11]:
from ultralytics import YOLO

model = YOLO("/content/runs/classify/train/weights/best.pt")

metrics = model.val(
    data="dataset",
    split="test"
)

print(metrics.top1)

Ultralytics 8.4.83 🚀 Python-3.12.13 torch-2.11.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLOv8n-cls summary (fused): 30 layers, 1,437,442 parameters, 0 gradients, 3.3 GFLOPs
train: /content/dataset/train... found 90 images in 2 classes ✅ 
val: /content/dataset/val... found 12 images in 2 classes ✅ 
test: /content/dataset/test... found 12 images in 2 classes ✅ 
test: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1652.5±903.5 MB/s, size: 80.8 KB)
test: Scanning /content/dataset/test... 12 images, 0 corrupt: 100% ━━━━━━━━━━━━ 12/12 5.6Mit/s 0.0s
               classes   top1_acc   top5_acc: 100% ━━━━━━━━━━━━ 1/1 1.2it/s 0.8s
                   all      0.917          1
Speed: 0.3ms preprocess, 1.0ms inference, 0.0ms loss, 0.0ms postprocess per image
Results saved to /content/runs/classify/val-3
0.9166666865348816


Sample test case on a single image


In [13]:
from ultralytics import YOLO

model = YOLO("/content/runs/classify/ScreenDetector/YOLOv8n/weights/best.pt")

result = model.predict(
    source="/content/57.jpg",
    imgsz=224,
    verbose=False
)[0]

# Predicted class
label = result.names[result.probs.top1]

# Confidence
confidence = float(result.probs.top1conf)

print("Predicted Label:", label)
print(f"Confidence: {confidence:.4f}")

Predicted Label: screen
Confidence: 0.9999


Downloading FILES

In [16]:
!zip -r ScreenDetector.zip /content/runs/classify/train

  adding: content/runs/classify/train/ (stored 0%)
  adding: content/runs/classify/train/weights/ (stored 0%)
  adding: content/runs/classify/train/weights/best.pt (deflated 8%)
  adding: content/runs/classify/train/weights/last.pt (deflated 8%)
  adding: content/runs/classify/train/confusion_matrix_normalized.png (deflated 37%)
  adding: content/runs/classify/train/train_batch2.jpg (deflated 4%)
  adding: content/runs/classify/train/train_batch1.jpg (deflated 4%)
  adding: content/runs/classify/train/val_batch0_labels.jpg (deflated 3%)
  adding: content/runs/classify/train/confusion_matrix.png (deflated 39%)
  adding: content/runs/classify/train/val_batch0_pred.jpg (deflated 3%)
  adding: content/runs/classify/train/train_batch0.jpg (deflated 3%)
  adding: content/runs/classify/train/args.yaml (deflated 53%)
  adding: content/runs/classify/train/results.png (deflated 9%)
  adding: content/runs/classify/train/results.csv (deflated 64%)


In [17]:
from google.colab import files

files.download("ScreenDetector.zip")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

-----------------------------------------------------------------------------

***Conclusion:***

Model- **YOLOv8n-cls (default)**              
Validation Accuracy: **100%**
Test Accuracy: **91.67%** (11/12 correct)



Model- **YOLOv8n-cls** (With data augmentation) Test                   Validation Accuracy-**92%**

-----------------------------------------------------------------------------

***Argument:***

Since the dataset was created by me it has only 57 images, we reached the accuracy of 100% on **training** and **validation** set but since the images were not enough only **1** image got misclassified in the test data which caused **92%** accuracy.

We tried data augmentation but it comes out be even worse.
We tried complex models like **YOLOv8n-cls** but it turns out to be more complex and overfits.


-----------------------------------------------------------------------------